src/test/scala/chiselTests/PrintableSpec.scala

```diff
@@ -47,88 +49,93 @@ class PrintableSpec extends AnyFlatSpec with Matchers {
         case _ => fail(s"Regex to process Printf should work on $str!")
       }
     }
-
     firrtl.split("\n").collect {
       case PrintfRegex(matched) =>
         val (str, args) = processBody(matched)
         Printf(str, args)
     }
   }
 
+
+  // Generates firrtl, gets Printfs
+  // Returns None if failed match; else calls the partial function which could have its own check
+  // Returns Some(true) to caller 
+  // Not calling fail() here - and letting caller do so - helps in localizing errors correctly. 
+  private def generateAndCheck(gen : => RawModule)(check : PartialFunction[Seq[Printf],Unit]) = {
```

Comment:
Rather than returning an option and having the caller do `.getOrElse(fail())` could we instead return a `Boolean` and have the caller `assert` it?

Reply:
Good point - let me do that.

Reply:
@jackkoenig  - upon converting this function to return boolean and then using an assert at the end of every test - getting this long and cryptic error message - ideally we want the failure msg to be as simple as "test failed at PrintableSpec.scala [line number]

```
  PrintableSpec.this.generateAndCheck(new MyModule())(({
    @SerialVersionUID(value = 0) final <synthetic> class $anonfun extends scala.runtime.AbstractPartialFunction[Seq[PrintableSpec.this.Printf],Unit] with java.io.Serializable {
      def <init>(): <$anon: Seq[PrintableSpec.this.Printf] => Unit> = {
        $anonfun.super.<init>();
        ()
      };
      final override def applyOrElse[A1 <: Seq[PrintableSpec.this.Printf], B1 >: Unit](x1: A1, default: A1 => B1): B1 = ((x1.asInstanceOf[Seq[PrintableSpec.this.Printf]]: Seq[PrintableSpec.this.Printf]): Seq[PrintableSpec.this.Printf] @unchecked) match {
        case scala.`package`.Seq.unapplySeq[PrintableSpec.this.Printf](<unapply-selector>) <unapply> ((str: String, args: Seq[String]): PrintableSpec.this.Printf("An exact string", scala.`package`.Seq.unapplySeq[String](<unapply-selector>) <unapply> ())) => ()
        case (defaultCase$ @ _) => default.apply(x1)
      };
      final def isDefinedAt(x1: Seq[PrintableSpec.this.Printf]): Boolean = ((x1.asInstanceOf[Seq[PrintableSpec.this.Printf]]: Seq[PrintableSpec.this.Printf]): Seq[PrintableSpec.this.Printf] @unchecked) match {
        case scala.`package`.Seq.unapplySeq[PrintableSpec.this.Printf](<unapply-selector>) <unapply> ((str: String, args: Seq[String]): PrintableSpec.this.Printf("An exact string", scala.`package`.Seq.unapplySeq[String](<unapply-selector>) <unapply> ())) => true
        case (defaultCase$ @ _) => false
      }
    };
    new $anonfun()
  }: PartialFunction[Seq[PrintableSpec.this.Printf],Unit])) was false (PrintableSpec.scala:86)
```

Reply:
This is what I did in one the tests 

```
it should "pass exact strings through" in {
    class MyModule extends BasicTester {
      printf(p"An exact strings")
    }
    
    assert(generateAndCheck(new MyModule) {
      case Seq(Printf("An exact string", Seq())) =>
    }) 

```

Reply:
Or maybe I should just add some default message - wanted to avoid boiler plate stuff.

Reply:
I think there's a better way. The way `fail()` gives a line number is that it takes an `implicit pos: Position` argument, try adding that to `generateAndCheck`

Reply:
`org.scalactic.source.Position` I think is the complete path. It's very similar to Chisel's own `SourceInfo` 😂

Reply:
Slight tangent - why did we have our own SourceInfo as opposed to reusing Position? - ok topic for an offline discussion!

Reply:
btw @jackkoenig  you rock! Thanks for the awesome idea - it does seem to work! Now this generateAndCheck is truly self-contained! Will commit my changes upon further testing...